sw/airborne/modules/imu/imu.c

```diff
@@ -31,107 +31,247 @@
 #include "modules/imu/imu.h"
 #include "state.h"
 #include "modules/core/abi.h"
+#include "modules/energy/electrical.h"
+
+/** By default disable IMU integration calculations */
+#ifndef IMU_INTEGRATION
+#define IMU_INTEGRATION false
+#endif
+
+/** Default gyro calibration is empty */
+#ifndef IMU_GYRO_CALIB
+#define IMU_GYRO_CALIB {}
+#endif
+
+/** Default accel calibration is empty */
+#ifndef IMU_ACCEL_CALIB
+#define IMU_ACCEL_CALIB {}
+#endif
+
+/** Default mag calibration is empty */
+#ifndef IMU_MAG_CALIB
+#define IMU_MAG_CALIB {}
+#endif
+
+/** Default body to imu is 0 (radians) */
+#if !defined(IMU_BODY_TO_IMU_PHI) && !defined(IMU_BODY_TO_IMU_THETA) && !defined(IMU_BODY_TO_IMU_PSI)
+#define IMU_BODY_TO_IMU_PHI   0
+#define IMU_BODY_TO_IMU_THETA 0
+#define IMU_BODY_TO_IMU_PSI   0
+#endif
+PRINT_CONFIG_VAR(IMU_BODY_TO_IMU_PHI)
+PRINT_CONFIG_VAR(IMU_BODY_TO_IMU_THETA)
+PRINT_CONFIG_VAR(IMU_BODY_TO_IMU_PSI)
 
-#ifdef IMU_POWER_GPIO
-#include "mcu_periph/gpio.h"
+/* Detect depricated gyro configurations */
+#if defined(IMU_GYRO_P_NEUTRAL) || defined(IMU_GYRO_Q_NEUTRAL) || defined(IMU_GYRO_R_NEUTRAL) || defined(IMU_GYRO_P_SENS) ||  defined(IMU_GYRO_Q_SENS) || defined(IMU_GYRO_R_SENS)
+#error "Using older gyro calibration `IMU_GYRO_*` defines please replace with `IMU_GYRO_CALIB`!"
```

Comment:
I think that should be possible to do and is not that hard to implement

Reply:
I have implemented a solution for this that seems to work so far: https://github.com/enacuavlab/paparazzi/commit/938f9982f074be80f3225bf3f9ceda2cbb981d58 (`multi_imu` branch in enac repo).
It does have some side effect on the `init` and `set_config` functions, but I think this is fine.

Reply:
Thanks, looks good! Only question why 'imu_set_default_config will override user conf if params are not NULL' this was chosen? Because the airframe file should be able to overwrite(preferably calibrated) values, if I'm correct.

Reply:
I see that for the rotation this was the case. But my idea was there to have another bool like `rotation_calibrated`, since most of the time you don't want to calibrate this and use the defaults. For external sensors you can then manually set the rotation, which should be directly combined in initialization with the `BODY_TO_IMU`.

Reply:
The problem is that there is always a calibration, either from user or from default parameters, so the sensor is tagged as calibrated, even if only neutral or only scale is provided. What we won't is to set the defaults for the parts that are not calibrated yet. So I think you are right that my implementation is not correct, but the initial one is not working either. I think the correct way is to change the `calibrated` bool for a bitmask telling if neutral, scale or rotation are calibrated or not.